pkg/version/version.go

```diff
@@ -18,14 +18,28 @@ import (
 
 // DO NOT EDIT
 // This var is updated automatically as part of the build process
-var Version = "undefined"
+var Version = undefinedVersion
 
 const (
+	undefinedVersion             = "undefined"
 	VersionSubsystemName         = "linkerd-version"
 	CliCheckDescription          = "cli is up-to-date"
 	ControlPlaneCheckDescription = "control plane is up-to-date"
 )
 
+func init() {
+	// Use `$LINKERD_VERSION` as the version only if the version wasn't set at
+	// link time to minimize the chance of using it unintentionally. This
+	// mechanism allows the version to be bound at container build time instead
+	// of at execurtable link time to improve incremental rebuild efficiency.
+	if Version == undefinedVersion {
+		override := os.Getenv("LINKERD_VERSION")
```

Comment:
I'm a bit wary of allowing a binary to override its version via an environment variable; version information feels like something that should be engrained in the binary itself. That said, I agree that this will speed up repeated builds of the controller and web images, so I'm ok with the tradeoff.

I do think we should rename the variable to indicate that it's only intended to be used when the binary is running inside a docker container. Maybe we could call it `LINKERD_CONTAINER_VERSION_OVERRIDE` or something equally obscure. `LINKERD_VERSION` feels like it could be set accidentally and cause problems.

Reply:
OK, I'l rename to `LINKERD_CONTAINER_VERSION_OVERRIDE`.

> LINKERD_VERSION feels like it could be set accidentally and cause problems.

Note that the PR already provides protection against accidental overriding because it only uses the override if it would otherwise return "undefined"; i.e. it only uses the override when the executables weren't linked with the version number added.